### PR TITLE
Fully import legacy Join and GroupJoin tests

### PR DESCRIPTION
**Pull request by JonHanna**
Convert Join GroupJoin tests imported in #2874 into a more directly XUnit-ian
style.

JoinTests.Join9.Test9 was a duplicate of Join8.Test8 and the case its comment
suggests it was meant to cover was covered by Join6.Test6, so removed.

### CODE REVIEW

**Comment by JonHanna** [comment]
@dotnet-bot test this please.

File: src/System.Linq/tests/GroupJoinTests.cs [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+            
+            public override int GetHashCode()
+            {
+                // Not great, but it'll serve.
+                return name.GetHashCode() ^ orderID.Length ^ (total.Length * 31);
+            }
+            
+            public bool Equals(JoinRec other)
+            {
+                if(!string.Equals(name, other.name))
```
Comment (stephentoub) [inline src/System.Linq/tests/GroupJoinTests.cs]: Nit: throughout the code, please ensure there are spaces between keywords like ```if``` and ```for``` and subsequent parens.

File: src/System.Linq/tests/JoinTests.cs [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+            AnagramRec[] inner = new []
+            {
+                new AnagramRec{ name = "miT", orderID = 43455, total = 10 },
+                new AnagramRec{ name = "Prakash", orderID = 323232, total = 9 }
+            };
+            JoinRec[] expected = new []
+            {
+                new JoinRec{ name = "Prakash", orderID = 323232, total = 9 },
+                new JoinRec{ name = "Tim", orderID = 43455, total = 10 }
+            };
```
Comment (stephentoub) [inline src/System.Linq/tests/JoinTests.cs]: Where is expected used?

File: src/System.Linq/tests/JoinTests.cs [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
```
Comment (stephentoub) [inline src/System.Linq/tests/JoinTests.cs]: Ditto

**Comment by stephentoub** [comment]
A few nits, but otherwise LGTM.

**Comment by JonHanna** [comment]
Nits picked.

**Pull request merged**